### DISCUSSION TITLE
Where in code are hitObjects displayed?

### DISCUSSION THREAD

**Discussion by Personator01** [ppy/osu#13624] [Q&A] [answered]
I've been sifting through the codebase for around two hours trying to find what code displays each hit circle as it appears on screen. As I understand it, DrawableHitObjects are added to the playfield all at once when the beatmap loads, but I have no clue when they are actually displayed.

**Comment by peppy** [comment]
please state your intentions so we can better help you.

**Reply by Personator01** [reply]
I'm trying to make for connectivity between osu and another piece of software (BCI2000 at bci2000.org), for research purposes, which would involve indicating when a hit circle or other target appears on screen.

**Reply by smoogipoo** [reply]
More specifically - are you looking to figure out how hitobjects are placed on the screen (position, size, size of approach circle, etc) so that you can create your own representations to feed to the software, or are you looking to extract a raw final representation of the hitobjects?

If the former, then the wiki post below provides some guidance to understand visuals along with reading/understanding `HitCircle`/`Slider`/`OsuHitObject`/`HitObject` "model" classes to better understand the specific values.

If the latter, then that is significantly more complex.

**Reply by Personator01** [reply]
It is the former, I just need to run some code whenever a hitobject appears, as communication to the other software is just sending some number over udp.

**Comment by frenzibyte** [comment, marked as answer]
I've written a wiki page explaining the system behind this, you can read it here: https://github.com/ppy/osu/wiki/Hitobject-drawables. You'll require knowledge on `Drawable` lifetime and transforms to understand that, however.

**Reply by Personator01** [reply]
Thanks a ton, I didn't see it when I first went over the wiki, sorry for the useless question

**Reply by frenzibyte** [reply]
It didn't exist previously, I've written it just now so no problem on that. :p

**Reply by Personator01** [reply]
As a follow up, are `UpdateStartTimeStateTransforms` or `UpdateState` called outside of `DrawableHitObject`, as in, when does the game itself update the hit objects?

**Reply by frenzibyte** [reply]
You can look for that yourself, find all places where `updateState` is called, check from where are they called, etc.
